File: frontend/src/app/features/tickets/services/ticket.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { environment } from '../../../../environments/environment';
import { Observable } from 'rxjs';

export interface TicketComment {
    id: string;
    ticketId: string;
    userId: string;
    userName?: string;
    commentText: string;
    isInternal: boolean;
    createdAt: string;
}

export interface Ticket {
    id: string;
    ticketNumber: string;
    title: string;
    description: string;
    status: string;
    priority: string;
    customerId?: string;
    customerName?: string;
    assignedToUserId?: string;
    assignedToName?: string;
    createdByUserId: string;
    createdByName?: string;
    createdAt: string;
    updatedAt: string;
    comments: TicketComment[];
}

export interface CreateTicketDto {
    title: string;
    description: string;
    priority: string;
    customerId?: string;
    assignedToUserId?: string;
}

export interface UpdateTicketDto {
    title: string;
    description: string;
    status: string;
    priority: string;
    customerId?: string;
    assignedToUserId?: string;
}

export interface CreateTicketCommentDto {
    commentText: string;
    isInternal: boolean;
}

@Injectable({ providedIn: 'root' })
export class TicketService {
    private apiUrl = `${environment.apiUrl}/tickets`;

    constructor(private http: HttpClient) { }

    getAll(page = 1, pageSize = 100): Observable<Ticket[]> {
        return this.http.get<Ticket[]>(`${this.apiUrl}?pageNumber=${page}&pageSize=${pageSize}`);
    }

    getById(id: string): Observable<Ticket> {
        return this.http.get<Ticket>(`${this.apiUrl}/${id}`);
    }

    create(dto: CreateTicketDto): Observable<Ticket> {
        return this.http.post<Ticket>(this.apiUrl, dto);
    }

    update(id: string, dto: UpdateTicketDto): Observable<void> {
        return this.http.put<void>(`${this.apiUrl}/${id}`, dto);
    }

    delete(id: string): Observable<void> {
        return this.http.delete<void>(`${this.apiUrl}/${id}`);
    }

    addComment(id: string, dto: CreateTicketCommentDto): Observable<TicketComment> {
        return this.http.post<TicketComment>(`${this.apiUrl}/${id}/comments`, dto);
    }
}

File: backend/CRM.Solution/src/CRM.Infrastructure/Services/TicketService.cs
using CRM.Application.Common.Exceptions;
using CRM.Application.Common.Interfaces;
using CRM.Application.Features.Tickets.DTOs;
using CRM.Domain.Entities;
using CRM.Infrastructure.Data;
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace CRM.Infrastructure.Services
{
    public class TicketService : ITicketService
    {
        private readonly ApplicationDbContext _context;

        public TicketService(ApplicationDbContext context)
        {
            _context = context;
        }

        public async Task<TicketDto?> GetTicketByIdAsync(Guid id)
        {
            var ticket = await _context.Tickets
                .Include(t => t.Customer)
                .Include(t => t.AssignedToUser)
                .Include(t => t.CreatedByUser)
                .Include(t => t.Comments)
                    .ThenInclude(c => c.User)
                .FirstOrDefaultAsync(t => t.Id == id);

            if (ticket == null) return null;
            return MapToDto(ticket);
        }

        public async Task<IReadOnlyList<TicketDto>> GetPagedTicketsAsync(int pageNumber, int pageSize)
        {
            var tickets = await _context.Tickets
                .Include(t => t.Customer)
                .Include(t => t.AssignedToUser)
                .Include(t => t.CreatedByUser)
                .OrderByDescending(t => t.CreatedAt)
                .Skip((pageNumber - 1) * pageSize)
                .Take(pageSize)
                .ToListAsync();

            return tickets.Select(MapToDto).ToList();
        }

        public async Task<IReadOnlyList<TicketDto>> GetPagedTicketsByUserAsync(Guid userId, int pageNumber, int pageSize)
        {
            var tickets = await _context.Tickets
                .Include(t => t.Customer)
                .Include(t => t.AssignedToUser)
                .Include(t => t.CreatedByUser)
                .Where(t => t.AssignedToUserId == userId || t.CreatedByUserId == userId)
                .OrderByDescending(t => t.CreatedAt)
                .Skip((pageNumber - 1) * pageSize)
                .Take(pageSize)
                .ToListAsync();

            return tickets.Select(MapToDto).ToList();
        }

        public async Task<TicketDto> CreateTicketAsync(CreateTicketDto createDto, Guid currentUserId)
        {
            var ticket = new Ticket
            {
                TicketNumber = $"TKT-{DateTime.UtcNow:yyyyMMdd}-{new Random().Next(1000, 9999)}",
                Title = createDto.Title,
                Description = createDto.Description,
                Priority = string.IsNullOrEmpty(createDto.Priority) ? "Medium" : createDto.Priority,
                Status = "Open",
                CustomerId = createDto.CustomerId,
                AssignedToUserId = createDto.AssignedToUserId,
                CreatedByUserId = currentUserId,
                CreatedAt = DateTime.UtcNow,
                UpdatedAt = DateTime.UtcNow
            };

            _context.Tickets.Add(ticket);
            await _context.SaveChangesAsync();

            return await GetTicketByIdAsync(ticket.Id) ?? MapToDto(ticket);
        }

        public async Task UpdateTicketAsync(Guid id, UpdateTicketDto updateDto, Guid currentUserId)
        {
            var ticket = await _context.Tickets.FindAsync(id);
            if (ticket == null) throw new NotFoundException("Ticket not found");

            ticket.Title = updateDto.Title;
            ticket.Description = updateDto.Description;
            ticket.Status = updateDto.Status;
            ticket.Priority = updateDto.Priority;
            ticket.CustomerId = updateDto.CustomerId;
            ticket.AssignedToUserId = updateDto.AssignedToUserId;
            ticket.UpdatedAt = DateTime.UtcNow;

            await _context.SaveChangesAsync();
        }

        public async Task<TicketCommentDto> AddCommentAsync(Guid ticketId, CreateTicketCommentDto commentDto, Guid currentUserId)
        {
            var ticket = await _context.Tickets.FindAsync(ticketId);
            if (ticket == null) throw new NotFoundException("Ticket not found");

            var comment = new TicketComment
            {
                TicketId = ticketId,
                UserId = currentUserId,
                CommentText = commentDto.CommentText,
                IsInternal = commentDto.IsInternal,
                CreatedAt = DateTime.UtcNow
            };

            _context.TicketComments.Add(comment);
            await _context.SaveChangesAsync();

            var user = await _context.Users.FindAsync(currentUserId);

            return new TicketCommentDto
            {
                Id = comment.Id,
                TicketId = comment.TicketId,
                UserId = comment.UserId,
                UserName = user != null ? $"{user.FirstName} {user.LastName}" : null,
                CommentText = comment.CommentText,
                IsInternal = comment.IsInternal,
                CreatedAt = comment.CreatedAt
            };
        }

        public async Task DeleteTicketAsync(Guid id)
        {
            var ticket = await _context.Tickets.FindAsync(id);
            if (ticket != null)
            {
                _context.Tickets.Remove(ticket);
                await _context.SaveChangesAsync();
            }
        }

        private TicketDto MapToDto(Ticket ticket)
        {
            return new TicketDto
            {
                Id = ticket.Id,
                TicketNumber = ticket.TicketNumber,
                Title = ticket.Title,
                Description = ticket.Description,
                Status = ticket.Status,
                Priority = ticket.Priority,
                CustomerId = ticket.CustomerId,
                CustomerName = ticket.Customer != null ? $"{ticket.Customer.FirstName} {ticket.Customer.LastName}" : null,
                AssignedToUserId = ticket.AssignedToUserId,
                AssignedToName = ticket.AssignedToUser != null ? $"{ticket.AssignedToUser.FirstName} {ticket.AssignedToUser.LastName}" : null,
                CreatedByUserId = ticket.CreatedByUserId,
                CreatedByName = ticket.CreatedByUser != null ? $"{ticket.CreatedByUser.FirstName} {ticket.CreatedByUser.LastName}" : null,
                CreatedAt = ticket.CreatedAt,
                UpdatedAt = ticket.UpdatedAt,
                Comments = ticket.Comments.Select(c => new TicketCommentDto
                {
                    Id = c.Id,
                    TicketId = c.TicketId,
                    UserId = c.UserId,
                    UserName = c.User != null ? $"{c.User.FirstName} {c.User.LastName}" : null,
                    CommentText = c.CommentText,
                    IsInternal = c.IsInternal,
                    CreatedAt = c.CreatedAt
                }).ToList()
            };
        }
    }
}
